transit.md

```diff
@@ -167,23 +170,167 @@ The Transit client can attempt connections to multiple relays, and uses the
 first one that passes negotiation. Each side combines a locally-configured
 hostname/port (usually "baked in" to the application, and hosted by the
 application author) with additional hostname/port pairs that come from the
-peer. This way either side can suggest the relays to use. The `wormhole`
-application accepts a `--transit-helper tcp:myrelay.example.org:12345`
-command-line option to supply an additional relay. The connection hints
+peer. This way either side can suggest the relays to use. The connection hints
 provided by the Transit instance include the locally-configured relay, along
 with the dynamically-determined direct hints. Both should be delivered to the
 peer.
 
+
+## Canonical abilities encodings
+
+The transit protocol relies on an existing secured (but possibly low-bandwidth)
+communication channel to exchange the abilities and hints. Thus, it has no
+influence over how they are encoded. However, we make a suggestion using JSON
+messages for other protocols to use.
+
+Abilities are encoded as a list, each item having a `type` tag. An ability may
+not appear more than once in the list. Setting `relay-v2` requires also setting
+`relay-v1` for backwards compatibility.
+
+```json
+[
+  {
+    "type": "<string, one of {direct-tcp-v1, relay-v1, tor-tcp-v1}>"
+  }
+]
+```
+
+Example for the full abilities set:
+
+```json
+[
+  { "type": "direct-tcp-v1" },
+  { "type": "relay-v1" },
+  { "type": "relay-v2" },
+  { "type": "tor-tcp-v1" }
+]
+```
+
+## Canonical hint encodings
+
+Hints are encoded as list of objects. Every object contains a `type` field,
+which further determines its encoding.
+
+### `direct-tcp-v1`, `tor-tcp-v1`
+
+```json
+{
+  "type": "direct-tcp-v1" or "tor-tcp-v1",
+  "hostname": <string>,
+  "port": <u16>,
+  "priority": <float; optional>
+}
+```
+
+`hostname` must be compliant with the `host` part of an URL, i.e. it may be an
+IP address or a domain. Furthermore, IPv6 link-local addresses are not supported.
+
+### `relay-v1`, `relay-v2`
+
+```json
+{
+  "type": "relay-v1",
+  "hints": [
+    {
+      "hostname": "<string>",
+      "port": "<tcp port>",
+      "priority": "<number, usually [0..1], optional>"
+    },
+    …
+  ],
+}
+```
+
+```json
+{
+  "type": "relay-v2",
+  "name": "<string, optional>",
+  "hints": [
```

Comment:
I thought I'd implement the proposed changes before merging (https://github.com/magic-wormhole/magic-wormhole.rs/pull/138/commits/d2771f28ae8a0293d26fe5d6f7f629b79e329a88). During this, I got a bit confused about the naming.

I think renaming `hints` to `endpoints` or something would make things less confusing.

Reply:
Okay I tried it out in the code, and it was a marginal improvement. However, while trying to writing it in the protocol I found that it would make the text more complicated. So in the end, it's probably not worth it.